Section on uninitialized memory (#5)
google/learn_unsafe_rust: src/advanced_unsafety/uninitialized.md

# Uninitialized memory

> _"I'm Nobody! Who are you? Are you — Nobody — too?"_
>
> — _Emily Dickinson_

While we have covered [invalid values], there's another thing that behaves a lot like invalid values, but has nothing to do with actual bit patterns: Uninitialized memory.

> An easy way to think about uninitialized memory is that there's an additional value (often called `undef` using LLVM's term for it) that does not map to any concrete bit pattern, but can be introduced in abstract in various ways, and makes _most_ values invalid.

Reviewer: "introduced in abstract" => "introduced in the abstract machine", I think that's more precise

> An easy way to think about uninitialized memory is that there's an additional value (often called `undef` using LLVM's term for it) that does not map to any concrete bit pattern, but can be introduced in the abstract machine in various ways, and makes _most_ values invalid.

Reviewer 2: i really don't like connecting to LLVM's names for stuff here bc LLVM `undef` is the wrong mental model e.g. LLVM guarantees the following:

```ll
%undef = i32 undef
%undef_hi = and i32 %undef, i32 0x0000ffff
%zero = and i32 %undef_hi, i32 0xffff0000
%is_zero = icmp i32 %zero, i32 0
call void @llvm.assume(i1 %is_zero)
```

i.e. `undef` is tracked at the bit level, so the bitpattern of any non-poison value is a ternary vector.

Rust models uninitialized memory much closer to LLVM's `poison`, which behaves holistically, more like a NaN: any operation involving `poison` is poison. Thus, `undef - undef == 0` but `poison - poison == poison` (insofar that == means anything here, because `icmp poison, poison` is also poison).

(Hilariously, langref does not specify whether or not `call void @llvm.assume(i1 poison)` is UB. lmao)

Reviewer 2: something else to discuss is that Rust does _not_ model uninit the way it LLVM models `poison`, because from LLVM's perspective operations with `poison` are well-defined *except* when: `load`, `store`, or `call` from a `ptr poison`, `br` on a `i1 poison`, or dividing by `poison`.

this is crucially different from how C++ models uninit, because C++ models almost exactly like `poison`, e.g. the following is fine in C++ and extremely not allowed in Rust:

```c++
struct X { int x; }

X x, y{0};
X* p = &x;
X* q = &y;

*q = *p;  // Copy from lvalue to uninit via X::X(const X&), y.x is now uninit.
y.x = 0;  // y is initialized again.
int z = p->x;  // Equivalent to int z;
```

Author: Yeah I think I'll kill this bit from the text.

Informally, people do say u BB def a l poo t, but we don't need to perpetuate that

> If you explicitly wish to work with uninitialized and partially-initialized types, [`MaybeUninit<T>`] is a useful abstraction since it can be "initialized" with no overhead and then written to in parts.

Reviewer: I'm not sure I like the "initialized" in quotes here. Maybe "... since a `MaybeUninit` value can be created with no overhead..."?

Author: Changed it to say constructed

## Sources of uninitialized memory

### `mem::uninitialized()` and `MaybeUninit::assume_init()`

> [`mem::uninitialized()`] is a deprecated API that has a very tempting shape, it lets you do things like `let x = mem::uninitialized()` for cases when you want to construct the value in bits. It's basically _always_ UB to use, since it immediately sets `x` to uninitialized memory, which is UB.

Reviewer: > which is UB

I think this is a good opportunity to reinforce exactly _why_ it's UB. Something like:

- Uninitialized values are invalid values (with few exceptions, e.g. `MaybeUninit` and other unions)
- `mem::uninitialized()` returned an uninitialized `T`
- Returning a value "produces" it
- It is undefined behavior to "produce" an invalid value

Not sure the most prosey way to phrase that, but I think that seeing that logic end-to-end is good.

Use [`MaybeUninit<T>`] instead.

> It is still possible to create uninitialized memory using [`MaybeUninit::assume_init()`] if you have not, in fact, assured that things are initialized.

Reviewer: "uninitialized memory" => "uninitialized values"

### Padding

> Padding bytes in structs and enums are often but not always uninitialized. This means that treating a struct as a bag of bytes (by, say, treating `&Struct` as `&[u8; size_of::<Struct>()]` and reading from there) is UB even if you don't write invalid values to those bytes, since you are ginning up uninitialized `u8`s.

Reviewer: I think we can strengthen this statement to "Padding bytes in structs and enums are uninitialized unless proven otherwise". It may also be good to mention how any typed copies (e.g. moves) may set the padding bytes back to `undef`.

The bag of bytes statement is a really good example since I think almost everyone has held that misconception at some point.

"ginning up" => "accessing": if they're reading them then I think we can use a little more straightforward terminology. Maybe "producing" instead to match the terminology from the book better? This section seems like it was originally working with deep reference validity.

Author: Yeah so I wanted to bring this up: you've mentioned before that padding bytes can infact be "not uninit", but looking at the UCG glossary there, it seems like they are _always_ uninit.

I actually think this can be strengthened to saying that they're always uninitialized when accessed via the struct. And perhaps say that if you _started_ with initialized bytes you can get them back but you have to use the original buffer.

(i.e. is it safe to go from `&[u8]` to `&Foo` to `&[u8]`?)

Reviewer: My guess is that we're talking about the same semantics for padding bytes, but this is basically what I mean by "uninitialized unless proven otherwise": https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=262fc5de7ac226e24770f56e1972eafb
In that case, it's always sound to read the padding bytes of the specific `Foo` being passed around by ref. If you have some random `Foo` value then you can do the same thing as long as you don't move it around (albeit that's very hard to avoid).

Author: Yeah that's not in concert with what UCG says: https://github.com/rust-lang/unsafe-code-guidelines/blob/master/reference/src/glossary.md#padding

> Copying Pad ignores the source byte, and writes any value to the target byte. Or, equivalently (in terms of Abstract Machine behavior), copying Pad marks the target byte as uninitialized.

Reviewer: We're just creating a reference to some `Foo` though, we're not copying it around. And even if we did copy out of it, the typed copy operation doesn't modify the source bytes so padding would remain unchanged (initialized to zero).

Author: we copied bytes out of it though?

Reviewer: Yes but we copied `u8`s out, not `Pad`s.

Author: Given that there is no way to copy the unnameable Pad type out as itself, I fail to see how the UCG glossary entry makes any sense unless it is talking about cast-and-then-read

Reviewer: The way I read it was that `Pad` is only used in the struct definition, and so only performs its special copy behavior when the struct is moved (i.e. as part of the struct). Basically the struct type is valid to _overlay_ on top of the bytes, moving the value out won't modify those original padding bytes, and you can cast back to a `[u8]` and read them. If you copied/moved the value, then the new value _will_ have uninit padding.

Author: Ah, fair point. I think we should file an upstream issue with concrete examples.

Author: https://github.com/rust-lang/unsafe-code-guidelines/issues/395

Author: Added a little "we don't know what the details are" header

Author: I got my answer and have reworked the section

Reading from padding [always produces uninitialized values][pad-glossary].

### Moved-from values

The following code is UB:

```rust
# use std::ptr;
let x = String::new(); // String is not Copy
let mut v = vec![];
let ptr = &x as *const String;

v.push(x); // move x into the vector

unsafe {
    // reads from moved-from memory

>     let ghost = ptr::read(ptr);

Reviewer: Isn't the issue here that `ptr` is dangling after `x` is moved, not that the value at `ptr` is necessarily uninitialized?

Author: I feel like it's both, right? It can be analyzed through the framework of dangling pointers or through the framework of uninit.

Reviewer 3: I think it is misleading to talk about deallocated and uninitialized memory in the same way. Reads from deallocated memory are insta-UB, whereas reads from uninitialized memory might not be UB if you're reading it as a type such as `MaybeUninit<u8>`.

Author: Yeah I'm going to rewrite this section. I'll keep it in since it's _similar_ to uninitialized memory, so worth thinking about, but not the same.

Author: I've reworked this a bit, thoughts?

Reviewer 3: I would reword this to start by talking about freed allocations. Then end the section by saying that moving out of a local variable is like deallocating the memory because the compiler might reuse that location on the stack for another local.

Author: I'm hesitant to do _too much_ here because I suspect that coverage may live better in a separate chapter. I mentioned freed memory a bit, though.

}
```

Any type of move will do this, even when you "move" the value into a different variable with stuff like `let y = x;`.

> Note that Rust does let you "partially move" out of fields of a struct, in such a case the whole struct is now no longer a valid value for its type, but you are still allowed to "use" the struct to look at other fields. When doing such things, make sure there are no pointers that still think the struct is whole and valid.

Reviewer: Might be worth mentioning that performing a partial move from a value makes the value _as a whole_ inaccessible. For example:

```rust
let x = (Foo::new(), Foo::new());
// After this assignment, what we call `x.0` is not actually a field
// of `x` any more because `x` doesn't exist any more.
let y = x.0;
// We can access and move out of `x.1` here even though `x`
// is no longer a valid `(Foo, Foo)` because `x` doesn't exist.
// What we call `x.1` is really just some `Foo` that was part of `x`.
let z = x.1;
```

Which is why you can't pass `x` anywhere or take references to it after performing a partial move.

#### Caveat: `ptr::drop_in_place()`, `ManuallyDrop::drop()`, and `ptr::read()`

> [`ptr::drop_in_place()`] and [`ManuallyDrop::drop()`] are interesting: they call all the destructor[^1] on a value (or a pointed-to value in the case of the former). From a safety point of view they are identical; they are just different APIs for dealing with manually calling drop glue.

Reviewer: "they call all" => "they both call"
"the former" => "drop_in_place" could make the parenthesis clearer
"a safety point of view" => "the perspective of safety"/"a safety perspective" tightens up the prose a bit

Might be helpful to put a footnote on "drop glue" as well or split that footnote in two.

[`ManuallyDrop::drop()`] makes the following claim:

> Other than changes made by the destructor itself, the memory is left unchanged, and so as far as the compiler is concerned still holds a bit-pattern which is valid for the type T.

> In other words, Rust does _not_ consider these operations to do the same invalidation as a regular "move from" operation, even though they have a similar feel.

Reviewer: I'm getting tripped up by some terminology here. I'm pretty sure that the kind of "move from" operation mentioned here doesn't "leave anything behind" - not even memory to access. In a sense, the builtin "move from" operation is either:

- Renaming an existing value
- or performing a typed copy out of some stack-allocated value (the stack may no longer contain the memory for the value afterward)

Author: Hmm, how would you frame this?

Author: As I said earlier I don't think this is super different from sticking an uninitialized value in the original memory

Reviewer: I've always thought of `ptr::read` as the "move from" operation, with the additional freeing as a separate step that occurs right afterward. Maybe it could be viewed as "as soon as a memory location is no longer used, it's undef".

Author: Hmm, that makes sense. I'm heading up the hill to do some trailwork in a bit, but if you'd like feel free to merge this and propose alternate wording? Or we can leave this open and try and figure it out together.

Author: I've reworked this a bit, thoughts?

There is an [open issue][ugc-394] about whether `Drop::drop()` is itself allowed to produce uninitialized or invalid memory, so it may not be possible to rely on this in a generic context.

[`ptr::read()`] similarly claims that it leaves the source memory untouched, which means that it is still a valid value. Of course, [`ptr::read()`] on a pointer pointing to uninitialized memory will still create an uninitialized value.

For all of these APIs, actually _using_ the dropped or read-from memory may still be fraught depending on the invariants of the value; it's quite easy to cause a double-free by materializing an owned value from the original data after it has already been read-from or dropped.

However, they do not produce uninitialized memory.

### Freshly allocated memory

> Freshly allocated memory (e.g. the yet-unused bytes in [`Vec::with_capacity()`] or just the result of [`Allocator::allocate()`]) is usually uninitialized. You can use APIs like [`Allocator::zeroed()`] if you wish to avoid this, though you can still end up making [invalid values] the same way you can with [`mem::zeroed()`].

Reviewer: `Allocator::zeroed` => `Allocator::allocate_zeroed`

# Uninitialized memory

> _"I'm Nobody! Who are you? Are you — Nobody — too?"_
>
> — _Emily Dickinson_

While we have covered [invalid values], there's another thing that is a kind of invalid value, but has nothing to do with actual bit patterns: Uninitialized memory.

An easy way to think about uninitialized memory is that there's an additional value that does not map to any concrete bit pattern, but can be introduced in the abstract machine in various ways, and makes _most_ values invalid.

> If you explicitly wish to work with uninitialized and partially-initialized types, [`MaybeUninit<T>`] is a useful abstraction since it can be constructed with no overhead and then written to in parts.

Reviewer 3: I would probably explain uninitialized memory along these lines:

 1. Uninitialized memory behaves exactly like a 257th value for bytes.
 2. You cannot treat an uninitialized byte as "this is just a byte with one of the 256 normal values picked randomly". Any attempt to figure out "what the byte actually is" is UB.
 3. Some types have invalid values, and uninitialized bytes are not allowed for most types.
 4. The `MaybeUninit<T>` type allows any sequence of bytes, including uninitialized ones. The padding of structs also allows any sequence of bytes, including uninitialized ones.
 5. If you copy bytes from one place to another with `std::ptr::copy` (which is _untyped_), then this perfectly copies the bytes, including whether they are uninitialized.
 6. If you copy bytes using a _typed copy_, then this promises fewer things than an untyped copy: it will only preserve things relevant for that type. For one, it is UB if the source bytes are invalid for the type. For another, if the type is a struct, then padding is not copied — instead, uninitialized bytes are written to the destination even if the source has initialized the padding (of course, the compiler might _implement_ this by just copying the padding, but it also might not). 
 7. Memory is uninitialized when it comes from the allocator. When the memory is freed, this is worse than just being uninitialized — it is UB to read it even as an `MaybeUninit` or with an untyped copy.

Thoughts?

Author: I've expanded this a bunch, thanks!

# Uninitialized memory

> _"I'm Nobody! Who are you? Are you — Nobody — too?"_
>
> — _Emily Dickinson_

While we have covered [invalid values], there's another thing that is a kind of invalid value, but has nothing to do with actual bit patterns: Uninitialized memory.

An easy way to think about uninitialized memory is that there's an additional value that does not map to any concrete bit pattern, but can be introduced in the abstract machine in various ways, and makes _most_ values invalid.

If you explicitly wish to work with uninitialized and partially-initialized types, [`MaybeUninit<T>`] is a useful abstraction since it can be constructed with no overhead and then written to in parts.

> ## Safely working with uninitialized memory

Author: not super happy with the flow of the chapter here, I think it's important to talk about this but I'm not sure _where_ it should go. We have spent a lot of time talking about where uninitialized values come from, but not talking about what actually is safe vs unsafe about it.

Author: I'm happier with it now but I think "Safely working with uninit" is now not the right name for the section

# Uninitialized memory

> _"I'm Nobody! Who are you? Are you — Nobody — too?"_
>
> — _Emily Dickinson_

While we have covered [invalid values], there's another thing that is a kind of invalid value, but has nothing to do with actual bit patterns: Uninitialized memory.

## Safely working with uninitialized memory

> The basic rule of thumb is: never refer to uninitialized memory with anything other than a raw pointer something wrapped in [`MaybeUninit<T>`]. Having a stack value or temporary that is uninitialized and has a type that is not `MaybeUninit<T>`  (or an array of `MaybeUninit`s) is always undefined behavior.

Reviewer 3: ```suggestion
The basic rule of thumb is: never refer to uninitialized memory with anything other than a raw pointer or something wrapped in [`MaybeUninit<T>`]. Having a stack value or temporary that is uninitialized and has a type that is not `MaybeUninit<T>`  (or an array of `MaybeUninit`s) is always undefined behavior.
```

A good model for uninitialized memory is that there's an additional value that does not map to any concrete bit pattern (think of it as "byte value #257"), but can be introduced in the abstract machine in various ways, and makes _most_ values invalid.

> Any attempt to read this byte as a `u8` will be UB, and the presence of this byte in non-padding locations is considered UB for most types. The exceptions to this all fall out of treating it as a property of the byte:

Reviewer 3: ```suggestion
Any attempt to read uninitialized bytes as an integer will be UB, and the presence of this byte in non-padding locations is considered UB for most types. The exceptions to this all fall out of treating it as a property of the byte:
```

> Any attempt to read uninitialized bytes as an integer will be UB, and the presence of this byte in non-padding locations is considered UB for most types. The exceptions to this all fall out of treating it as a property of the byte:

Reviewer: ```suggestion
Any attempt to read uninitialized bytes will be UB, and the presence of this byte in non-padding locations is considered UB for most types. The exceptions to this all fall out of treating it as a property of the byte:
```

Author: The caveat was introduced because you can read it as a MaybeUninit. I'll try to make it clear that  it's "as a type that cares about initializedness"

 - Zero-sized types do not care about initialized-ness, since they do not have bytes
 - Unions do not care about initialized-ness if they have a variant that does not care about initialized-ness
 - [`MaybeUninit<T>`] does not care about initializedness since it is internally a union of `T` and a zero-sized type.
 - `[MaybeUninit<T>; N]` [does not care about initializedness][arr-maybeuninit] since it doesn't have any bytes that care about initializedness
 

> Fundamentally, initializedness is a property of memory, but whether or not initializedness matters is a property of the *type*. For types that care about initializedness, typed operations working with uninitialized memory are typically UB, and having a value that contains uninitialized memory is immediately UB.

Reviewer 3: ```suggestion
Fundamentally, initializedness is a property of memory, but whether or not initializedness matters is a property of the access (in particular, of the *type* used by the access). For types that care about initializedness, typed operations working with uninitialized memory are typically UB, and having a value that contains uninitialized memory is immediately UB.
```

[`ptr::copy`] is explicitly an *untyped* copy, and thus it will copy all bytes, including padding, and including initialized-ness, to the destination, regardless of the type `T`.

> Most other operations copying a type (for example, `*ptr` and `mem::transmute_copy`) will be typed, and will thus ignore padding and be UB if ever fed uninitialized memory in non-padding positions. This also applies to `let x = y` and `mem::transmute`, however in those cases if the source data were uninitialized that would already have been UB.

Reviewer 3: Not if you're transmuting from `MaybeUninit<u8>` to `u8`.

If you explicitly wish to work with uninitialized and partially-initialized types, [`MaybeUninit<T>`] is a useful abstraction since it can be constructed with no overhead and then written to in parts. It's also useful to e.g. refer to an uninitialized buffer with things like `&mut [MaybeUninit<u8>]`.

> Similarly with invalid values, there are open issues ([UGC #77], [UGC #346]) about whether it is UB to have _references_ to uninitialized memory. When writing unsafe code we recommend you avoid creating such references, choosing to always use `MaybeUninit`, but when auditing unsafe code there may be causes where a reference to uninitialized values is actually safe as long as no uninitialized value is read out of it. In particular, [UGC #346] indicates that it is extremely unlikely that having `&mut` references to uninitialized values will be immediately UB.

Reviewer: ```suggestion
Similarly with invalid values, there are open issues ([UGC #77], [UGC #346]) about whether it is UB to have _references_ to uninitialized memory. When writing unsafe code we recommend you avoid creating such references, choosing to always use `MaybeUninit`. When auditing unsafe code, there may be cases where a references to uninitialized values are actually safe as long as no uninitialized values are read out of it. In particular, [UGC #346] indicates that it is extremely unlikely that having `&mut` references to uninitialized values will be immediately UB.
```

## Sources of uninitialized memory

### `mem::uninitialized()` and `MaybeUninit::assume_init()`

> [`mem::uninitialized()`] is a deprecated API that has a very tempting shape, it lets you do things like `let x = mem::uninitialized()` for cases when you want to construct the value in bits. It's almost _always_ UB to use, since it immediately sets `x` to uninitialized memory, which is UB, since uninitialized memory is a type of invalid value for almost all types, and it's unsound to produce invalid values.

Reviewer: ```suggestion
[`mem::uninitialized()`] is a deprecated API that has a very tempting shape: it lets you do things like `let x = mem::uninitialized()` when you want to construct an uninitialized value. It's _almost always_ UB to use since it immediately sets `x` to uninitialized memory, which is UB because uninitialized memory is an invalid value for almost all types and it's unsound to produce invalid values.
```

Use [`MaybeUninit<T>`] instead.

> It is still possible to create uninitialized values using [`MaybeUninit::assume_init()`] if you have not, in fact, assured that things are initialized.

Reviewer 3: What about `MaybeUninit::uninit()`? The `assume_init` method is not the _source_ of the uninitialized bytes. I think its better to start by focusing `MaybeUninit` and ending it by saying "`mem::uninitialized()` is exactly equivalent to `MaybeUninit::uninit().assume_init()`, but it is deprecated in favor of the latter way of writing it"

### Padding

Padding bytes in structs and enums are [usually but not always uninitialized][pad-glossary]. This means that treating a struct as a bag of bytes (by, say, treating `&Struct` as `&[u8; size_of::<Struct>()]` and reading from there) is UB even if you don't write invalid values to those bytes, since you are accessing uninitialized `u8`s.

The "usually but not always" caveat can be usefully framed as "padding bytes are uninitialized unless proven otherwise". Padding is a property of types, not memory, and these bytes are set to being uninitialized whenever a type is created or copied/moved around, but they can be written to by getting a reference to the memory behind the type[^1], and will be preserved at that spot in memory as long as the type isn't overwritten as a whole.

> For example, treating an initialized byte buffer as an `&Struct` and then later reading the padding bytes will give initialized values. However, treating an initialized byte buffer as an `&mut Struct` and then writing a new `Struct` to it will lead to those bytes becoming uninitialized since the `Struct` copy will "copy" the uninitialized padding bytes. Similarly, using `mem::transmute()` (or `mem::zeroed()`) to transmute a byte buffer to a `Struct` will have the padding be uninitialized, because a typed copy of the `Struct` is occurring.

Reviewer: ```suggestion
For example, treating an initialized byte buffer as an `&Struct` and then later reading the padding bytes will give initialized values. However, treating an initialized byte buffer as an `&mut Struct` and then writing a new `Struct` to it will lead to those bytes becoming uninitialized since the `Struct` copy will "copy" the uninitialized padding bytes. Similarly, using `mem::transmute()` (or `mem::zeroed()`) to transmute a byte buffer to a `Struct` will uninitialize the padding because it performs a typed copy of the `Struct`.
```

Because [`ptr::copy`] is an untyped copy, it can be used to copy over explicitly-initialized padding.

> See the discussion in [UGC #395][ugc395] for more examples.

Reviewer 3: I like this part:

> The "usually but not always" caveat can be usefully framed as "padding bytes are uninitialized unless proven otherwise".

But otherwise, I think this is somewhat unclear. I would be more explicit about the memory vs access aspect.

### Unions

> Reading a union type as the wrong variant can lead to reading uninitialized memory, for example if the union was initialized to a smaller variant, or if the padding of the two variants doesn't overlap perfectly.

Reviewer: Is "padding" as an aggregate noun? These are the questions this project was made to resolve.

```suggestion
Reading a union type as the wrong variant can lead to reading uninitialized memory, for example if the union was initialized to a smaller variant, or if the padding of the two variants don't overlap perfectly.
```

> Rust does not have strict aliasing like C and C++: type punning with a union can be safe as long as the punning does not cause invalid or uninitialized values to show up on the other side.

Reviewer 3: ```suggestion
Rust does not have strict aliasing like C and C++: type punning with a union is safe as long as the corresponding transmute is safe.
```

> Other than changes made by the destructor itself, the memory is left unchanged, and so as far as the compiler is concerned still holds a bit-pattern which is valid for the type T.

In other words, Rust does _not_ consider these operations to do the same invalidation as a regular "move from" operation, even though they may have a similar feel. They do not create dangling pointers, and they do not themselves overwrite the memory with an uninitialized value.

There is an [open issue][ugc-394] about whether `Drop::drop()` is itself allowed to produce uninitialized or invalid memory, so it may not be possible to rely on this in a generic context.

[`ptr::read()`] similarly claims that it leaves the source memory untouched, which means that it is still a valid value. Of course, [`ptr::read()`] on a pointer pointing to uninitialized memory will still create an uninitialized value.

For all of these APIs, actually _using_ the dropped or read-from memory may still be fraught depending on the invariants of the value; it's quite easy to cause a double-free by materializing an owned value from the original data after it has already been read-from or dropped.

> However, they do not produce uninitialized memory.

Reviewer 3: The reason that `ManuallyDrop` makes this promise is so that calling `ManuallyDrop::drop` on the value in an `Option<ManuallyDrop<T>>` doesn't suddenly turn the option into a `None` by zeroing the pointer that the option is using for its discriminant.

I would say "always treat these as leaving the memory in an uninitialized state" and that you should never rely on them doing any more than that, other than the guarantee that niche-optimizations still work for `ManuallyDrop`.

An exception is Copy types where both are guaranteed no-ops.

Author: I agree that that is the right advice for unsafe code writers, but it's not an accurate description of what is happening. I'll try to expand the section a bit here.
